File: app/src/main/java/com/iamdsr/travel/PlanTrip/AddTripFragment.java
package com.iamdsr.travel.PlanTrip;

import android.app.DatePickerDialog;
import android.app.ProgressDialog;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;

import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.iamdsr.travel.AppLaunchSetup.LoginActivity;
import com.iamdsr.travel.Models.TripModel;
import com.iamdsr.travel.R;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.concurrent.TimeUnit;

public class AddTripFragment extends Fragment {

    //widgets
    private View view;
    private EditText mJourneyDate, mReturnDate, mAddTitle, mAddDesc, mWhereFrom, mWhereTo, mNumberOfPerson;
    private Button mAddTripBtn;
    private EditText clickedEditText;

    //Firebase
    private FirebaseAuth mAuth;
    private FirebaseFirestore firebaseFirestore;

    //Utils
    private String currUserID = FirebaseAuth.getInstance().getCurrentUser().getUid();
    private static final String TAG = "AddTripFragment";
    final Calendar myCalendar= Calendar.getInstance();

    public AddTripFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_add_trip, container, false);
        setUpWidgets();
        setupFirebase();
        setUpDateDialogs();

        mAddTripBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addTripToDB();
            }
        });


        return view;
    }

    private void setUpDateDialogs(){
        final DatePickerDialog.OnDateSetListener date = new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                myCalendar.set(Calendar.YEAR, year);
                myCalendar.set(Calendar.MONTH, monthOfYear);
                myCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                updateLabel(clickedEditText);
            }
        };
        mJourneyDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                clickedEditText = (EditText) view;
                new DatePickerDialog(getContext(), date, myCalendar
                        .get(Calendar.YEAR), myCalendar.get(Calendar.MONTH),
                        myCalendar.get(Calendar.DAY_OF_MONTH)).show();
            }
        });
        mReturnDate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                clickedEditText = (EditText) view;
                new DatePickerDialog(getContext(), date, myCalendar
                        .get(Calendar.YEAR), myCalendar.get(Calendar.MONTH),
                        myCalendar.get(Calendar.DAY_OF_MONTH)).show();
            }
        });
    }

    private void addTripToDB(){
        String title = mAddTitle.getText().toString().trim();
        String desc = mAddDesc.getText().toString().trim();
        String from = mWhereFrom.getText().toString().trim();
        String to = mWhereTo.getText().toString().trim();
        String jDate = mJourneyDate.getText().toString().trim();
        String rDate = mReturnDate.getText().toString().trim();
        String totalPerson = mNumberOfPerson.getText().toString().trim();
        DocumentReference ref = firebaseFirestore.collection("trips").document();
        final String newTripID = ref.getId();
        getDateDiff(jDate,rDate);
        if (!TextUtils.isEmpty(title) && !TextUtils.isEmpty(desc) && !TextUtils.isEmpty(from)
                && !TextUtils.isEmpty(to) && !TextUtils.isEmpty(jDate) && !TextUtils.isEmpty(rDate) && !TextUtils.isEmpty(totalPerson)){
            ProgressDialog progressDialog = new ProgressDialog(getContext());
            progressDialog.setMessage("Adding Trip");
            progressDialog.show();
            progressDialog.setCancelable(false);
            progressDialog.setCanceledOnTouchOutside(false);
            TripModel tripModel = new TripModel(newTripID,title,desc,jDate,rDate,from,to,currUserID,getDateDiff(jDate,rDate),Long.parseLong(totalPerson),getTimestamp());
            firebaseFirestore.collection("users")
                    .document(currUserID)
                    .collection("trips")
                    .document(newTripID)
                    .set(tripModel)
                    .addOnSuccessListener(new OnSuccessListener<Void>() {
                @Override
                public void onSuccess(Void unused) {
                    progressDialog.dismiss();
                    mAddTitle.setText(null);
                    mAddDesc.setText(null);
                    mWhereFrom.setText(null);
                    mWhereTo.setText(null);
                    mJourneyDate.setText(null);
                    mReturnDate.setText(null);
                    mNumberOfPerson.setText(null);
                    Navigation.findNavController(view).navigate(R.id.action_addTripFragment_to_planTripFragment);
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(@NonNull Exception e) {

                }
            });
        }
        else{

        }
    }

    private void updateLabel(EditText view) {
        String myFormat="dd/MM/yyyy";
        SimpleDateFormat dateFormat=new SimpleDateFormat(myFormat, Locale.getDefault());
        view.setText(dateFormat.format(myCalendar.getTime()));
    }

    private long getDateDiff(String CurrentDate, String FinalDate){
        try {
            Date date1;
            Date date2;
            String myFormat="dd/MM/yyyy";
            SimpleDateFormat dateFormat=new SimpleDateFormat(myFormat, Locale.getDefault());
            date1 = dateFormat.parse(FinalDate);
            date2 = dateFormat.parse(CurrentDate);
            long difference = Math.abs(date1.getTime() - date2.getTime());
            Log.d(TAG, "getDateDiff: currdate : "+date1);
            Log.d(TAG, "getDateDiff: enddate : "+date2);
            Log.d(TAG, "getDateDiff: difference : "+difference/(24 * 60 * 60 * 1000));
            return difference / (24 * 60 * 60 * 1000);
        } catch (Exception e) {
            Log.d(TAG, "getDateDiff: "+e.getMessage());
            return 0;
        }
    }

    private String getTimestamp(){
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
        format.setTimeZone(TimeZone.getDefault());
        return format.format(new Date());
    }

    private void setupFirebase(){
        mAuth = FirebaseAuth.getInstance();
        firebaseFirestore = FirebaseFirestore.getInstance();
    }

    private void setUpWidgets() {
        mAddTitle = view.findViewById(R.id.trip_title);
        mAddDesc = view.findViewById(R.id.trip_desc);
        mWhereFrom = view.findViewById(R.id.place_from);
        mWhereTo = view.findViewById(R.id.place_to);
        mNumberOfPerson = view.findViewById(R.id.total_person);
        mJourneyDate = view.findViewById(R.id.journey_date);
        mReturnDate = view.findViewById(R.id.return_date);
        mAddTripBtn = view.findViewById(R.id.add_trip_btn);
    }
}

File: app/src/main/java/com/iamdsr/travel/planTrip/AddTripFragment.kt
package com.iamdsr.travel.planTrip

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.fragment.findNavController
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.iamdsr.travel.R
import com.iamdsr.travel.models.TripModel
import com.iamdsr.travel.repositories.FirestoreRepository
import com.iamdsr.travel.viewModels.PlanTripFragmentViewModel
import java.text.SimpleDateFormat
import java.util.*


class AddTripFragment : Fragment() {

    // Widgets
    private lateinit var mJourneyDate: TextInputEditText
    private lateinit var mReturnDate:TextInputEditText
    private lateinit var mAddTitle:TextInputEditText
    private lateinit var mAddDesc:TextInputEditText
    private lateinit var mWhereFrom:TextInputEditText
    private lateinit var mWhereTo:TextInputEditText
    private lateinit var mNumberOfPerson:TextInputEditText
    private lateinit var mJourneyMode: AutoCompleteTextView

    private lateinit var mJourneyDateContainer: TextInputLayout
    private lateinit var mReturnDateContainer:TextInputLayout
    private lateinit var mAddTitleContainer:TextInputLayout
    private lateinit var mAddDescContainer:TextInputLayout
    private lateinit var mWhereFromContainer:TextInputLayout
    private lateinit var mWhereToContainer:TextInputLayout
    private lateinit var mNumberOfPersonContainer:TextInputLayout
    private lateinit var mJourneyModeContainer: TextInputLayout

    private lateinit var snack_layout: CoordinatorLayout

    private var mAddTripBtn: Button? = null
    private var progressBar: ProgressBar? = null

    // Utils
    private val journeyModeEntries = arrayOf("Flight", "Train", "Car")

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_add_trip, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpWidgets()
        mJourneyDate.setOnClickListener(View.OnClickListener {
            setUpDateDialogs(it as TextInputEditText)
        })
        mReturnDate.setOnClickListener(View.OnClickListener {
            setUpDateDialogs(it as TextInputEditText)
        })
        mAddTripBtn!!.setOnClickListener {
            addNewTrip()
        }
    }


    private fun addNewTrip() {
        if (FirebaseAuth.getInstance().currentUser!=null) {
            val title = mAddTitle.text.toString().trim()
            val desc = mAddDesc.text.toString().trim()
            val from = mWhereFrom.text.toString().trim()
            val to = mWhereTo.text.toString().trim()
            val jDate = mJourneyDate.text.toString().trim()
            val rDate = mReturnDate.text.toString().trim()
            val totalPerson = mNumberOfPerson.text.toString().trim()
            val journeyMode = mJourneyMode.text.toString().trim()



            val planTripFragmentViewModel = ViewModelProvider(this)[PlanTripFragmentViewModel::class.java]
            if ((!TextUtils.isEmpty(title) && !TextUtils.isEmpty(desc) && !TextUtils.isEmpty(from)
                        && !TextUtils.isEmpty(to) && !TextUtils.isEmpty(jDate) && !TextUtils.isEmpty(rDate) &&
                        !TextUtils.isEmpty(totalPerson) && !TextUtils.isEmpty(journeyMode))) {

                val error = checkDataOfInputFields(title, desc, from, to, jDate, rDate, totalPerson, journeyMode)
                Log.d("TAG", "addNewTrip: Error $error")

                if (!error){
                    progressBar?.visibility = View.VISIBLE
                    val firebaseRepository = FirestoreRepository()
                    val tripModel = TripModel(
                        firebaseRepository.getNewTripID(),
                        title,
                        desc,
                        jDate,
                        rDate,
                        from,
                        to,
                        FirebaseAuth.getInstance().currentUser!!.uid,
                        getTimestamp(),
                        journeyMode,
                        getDateDiff(jDate, rDate),
                        totalPerson.toLong()
                    )
                    planTripFragmentViewModel._addNewTripToFirebaseFirestore(tripModel)
                    findNavController().navigate(R.id.action_addTripFragment_to_planTripFragment)
                    progressBar?.visibility = View.INVISIBLE
                }
            }
            else {
                Snackbar.make(snack_layout, context!!.resources.getString(R.string.empty_fields_msg), Snackbar.LENGTH_LONG).show()
            }
        }
    }
    private fun checkDataOfInputFields(title: String, desc: String, from: String, to: String, jDate: String, rDate: String, pax: String, jMode: String): Boolean{
        var errorFound = false
        if (title.length > 75){
            errorFound = true
            mAddTitleContainer.error = context!!.getString(R.string.length_exceeds_error, 75)
        }
        else {
            mAddTitleContainer.error = null
        }
        if (desc.length > 300){
            errorFound = true
            mAddDescContainer.error = context!!.getString(R.string.length_exceeds_error, 300)
        }
        else {
            mAddDescContainer.error = null
        }
        if (checkDates(jDate, rDate) != ""){
            errorFound = true
            mJourneyDateContainer.error = checkDates(jDate, rDate)
            mReturnDateContainer.error = checkDates(jDate, rDate)
        }
        else {
            mJourneyDateContainer.error = null
            mReturnDateContainer.error = null
        }
        if (from.length > 20){
            errorFound = true
            mWhereFromContainer.error = context!!.getString(R.string.length_exceeds_error, 20)
        }
        else {
            mWhereFromContainer.error = null
        }
        if (to.length > 20){
            errorFound = true
            mWhereToContainer.error = context!!.getString(R.string.length_exceeds_error, 20)
        }
        else {
            mWhereToContainer.error = null
        }
        if (pax.toInt() < 0){
            errorFound = true
            mNumberOfPersonContainer.error = context!!.getString(R.string.invalid_value_error, pax)
        }
        else {
            mNumberOfPersonContainer.error = null
        }
        if (!journeyModeEntries.contains(jMode)){
            Log.d("TAG", "addNewTrip: Error $jMode")
            errorFound = true
            mJourneyModeContainer.error = context!!.getString(R.string.invalid_value_error, jMode)
        }
        else {
            mJourneyModeContainer.error = null
        }
        return errorFound
    }
    private fun checkDates(startDate: String, endDate: String): String? {
        val sdf = SimpleDateFormat("dd/mm/yyyy", Locale.getDefault())
        try {
            when {
                sdf.parse(endDate)!!.before(sdf.parse(startDate)) -> {
                    return context!!.resources.getString(R.string.end_date_l_start_date)
                }
                sdf.parse(endDate)!!.after(sdf.parse(startDate)) -> {
                    return ""
                }
                sdf.parse(endDate)!! == (sdf.parse(startDate)) -> {
                    return context!!.resources.getString(R.string.end_date_e_start_date)
                }
            }
        }
        catch (e: Exception){
            e.printStackTrace()
        }
        return null
    }
    private fun getTimestamp(): String {
        val format = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        format.timeZone = TimeZone.getDefault()
        return format.format(Date())
    }
    private fun setUpDateDialogs(view: TextInputEditText) {

        val datePicker = MaterialDatePicker.Builder.datePicker().setTheme(R.style.ThemeOverlay_App_DatePicker).build()
        datePicker.show(parentFragmentManager, "DatePicker")

        datePicker.addOnPositiveButtonClickListener {
            val dateFormatter = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
            val date = dateFormatter.format(Date(it))
            view.setText(date)
        }
        datePicker.addOnNegativeButtonClickListener {
            Toast.makeText(context, "${datePicker.headerText} is cancelled", Toast.LENGTH_LONG).show()
        }
        datePicker.addOnCancelListener {
            Toast.makeText(context, "Date Picker Cancelled", Toast.LENGTH_LONG).show()
        }


    }

    private fun getDateDiff(CurrentDate: String, FinalDate: String): Long {
        return try {
            val date1: Date
            val date2: Date
            val myFormat = "dd/MM/yyyy"
            val dateFormat = SimpleDateFormat(myFormat, Locale.getDefault())
            date1 = dateFormat.parse(FinalDate)
            date2 = dateFormat.parse(CurrentDate)
            val difference = Math.abs(date1.time - date2.time)
            difference / (24 * 60 * 60 * 1000)
        } catch (e: Exception) {
            0
        }
    }

    private fun setUpWidgets() {
        if (view != null){
            mAddTitle = view!!.findViewById(R.id.trip_title)
            mAddDesc = view!!.findViewById(R.id.trip_desc)
            mWhereFrom = view!!.findViewById(R.id.place_from)
            mWhereTo = view!!.findViewById(R.id.place_to)
            mNumberOfPerson = view!!.findViewById(R.id.total_person)
            mJourneyDate = view!!.findViewById(R.id.journey_date)
            mReturnDate = view!!.findViewById(R.id.return_date)
            mJourneyMode = view!!.findViewById(R.id.journey_mode)

            mAddTitleContainer = view!!.findViewById(R.id.trip_title_container)
            mAddDescContainer = view!!.findViewById(R.id.trip_desc_container)
            mWhereFromContainer = view!!.findViewById(R.id.place_from_container)
            mWhereToContainer = view!!.findViewById(R.id.place_to_container)
            mNumberOfPersonContainer = view!!.findViewById(R.id.total_person_container)
            mJourneyDateContainer = view!!.findViewById(R.id.journey_date_container)
            mReturnDateContainer = view!!.findViewById(R.id.return_date_container)
            mJourneyModeContainer = view!!.findViewById(R.id.journey_mode_container)

            snack_layout = view!!.findViewById(R.id.snackbar_layout)
            mAddTripBtn = view!!.findViewById(R.id.add_trip_btn)
            progressBar = view!!.findViewById(R.id.progress_horizontal)
            val journeyModeAdapter = ArrayAdapter(context!!, android.R.layout.simple_list_item_1, journeyModeEntries)
            mJourneyMode.setAdapter(journeyModeAdapter)
        }
    }
}